### ISSUE TITLE
gevent.threadpool  throws 'LoopExit: This operation would block forever' error and pool gives none thread under load test

### ISSUE THREAD

**Issue by vamkaxifa** [gevent/gevent#863]
* gevent version: 1.1.1
* Python version:2.7.9
* Operating System:Linux testint01 2.6.18-194.el5 #1 SMP Tue Mar 16 21:52:39 EDT 2010 x86_64 x86_64 x86_64 GNU/Linux

### Description:
I have a middleware which is between business system and oracle database system. This middleware operates database  for  business system. It uses gevent to start the server like this:
````
application = tornado.wsgi.WSGIApplication([
    (r"/healthcheck", HealthCheck)])
server = gevent.wsgi.WSGIServer(('', args.port), application, log=None)
server.serve_forever() 
````

Recently , I found that this middleware had performance bottlenecks under high load. It turuned out that cx_Oracle's OCI(Oracle Call Interface) doesn't provide asynchronous functions. reference: [https://bitbucket.org/anthony_tuininga/cx_oracle/issues/8/gevent-eventlet-support ]
Any request which spends longtime in oracle will be blocked, in this condition, the entire middleware blocked, all rest request can't recive  response  until oracle return. 
In order to solve this problem, I add threadpool to start a child thread to access database like this : 

```

class  BaseServices:

    def __init__(self, db_pool, wap_db_pool):
        self.wap_db_pool = wap_db_pool
        self._db_pool = db_pool
        self._db_handle_pool = ThreadPool(500)
        self._logger = logging.getLogger('default')
        self.db_ops = {}
        Logger.debug(self._logger, 'BaseDataQueryService created')

    def do_db_operation(self, op_id, conn, c, procedure_name, procedure_params):
        self._logger.debug('inside, parent thread is %s ,thread is %s' % (op_id, thread.get_ident()))
        try:
            p = c.callproc(procedure_name, procedure_params)
            self._logger.debug('inside after, parent thread is %s ,thread is %s, p is %s, db_pool_maxsize is %s, db_pool_size is %s'
                               % (op_id, thread.get_ident(), str(p), conn._pool._maxconnections, conn._pool._connections))

        except Exception as ex:
            self._logger.error(ex)
            msg = traceback.format_exc()
            self._logger.error('\n' + 'inside exception, parent thread is %s ,thread is %s'
                               % (op_id, thread.get_ident()) + ' ' + msg)
        self._logger.debug('inside end ,parent thread is %s , thread is %s, thread_pool_maxsize is %s, thread_pool_size is %s'
                           % (op_id, thread.get_ident(), str(self._db_handle_pool._get_maxsize()),
                              str(self._db_handle_pool._get_size())))
        # self.db_ops.get(op_id).set(p)
        return p

    def query_force_db(self, data):
        """
        force query db 
        :param data: post 
        :return:
        """
        procedure_id = None
        res_model = ResModel()
        try:
                ...   # parameter handling 
                conn, c = None, None
                try:
                    conn = self._db_pool.connection()
                    c = conn.cursor()
                    ...  # parameter handling
                    self._logger.debug('before, thread is %s , thread_pool_maxsize is %s, thread_pool_size is %s'
                                       % (thread.get_ident(), str(self._db_handle_pool._get_maxsize()),
                                          str(self._db_handle_pool._get_size())))
                    self._logger.debug('before, thread is %s, db_pool_maxsize is %s, db_pool_size is %s'
                                       % (thread.get_ident(), conn._pool._maxconnections, conn._pool._connections))
                    #   call procedure    procedure_name procedure_params   return p
                    p = self._db_handle_pool.apply(self.do_db_operation, (str(thread.get_ident()),conn,
                                                                          c, procedure_name, procedure_params))
                    # p = c.callproc(procedure_name, procedure_params)
                    self._logger.debug('after, p = %s ,thread = %s, thread_pool_maxsize is %s, thread_pool_size is %s'
                                       % (p, thread.get_ident(), str(self._db_handle_pool._get_maxsize()),
                                       str(self._db_handle_pool._get_size())))
                    self._logger.debug('after, thread = %s, db_pool_maxsize is %s, db_pool_size is %s'
                                       % (thread.get_ident(), conn._pool._maxconnections, conn._pool._connections))
                    for out_value in p[len(procedure_in_params):]:
                       ...  #  result handing 
                except Exception as ex:
                    self._logger.error(ex)
                    msg = traceback.format_exc()
                    self._logger.error('\n' + msg)
                    ...
                finally:
                    if conn is not None:
                        conn.close()
                    self._logger.debug('finally, thread = %s, db_pool_maxsize is %s, db_pool_size is %s'
                                       % (thread.get_ident(), conn._pool._maxconnections, conn._pool._connections))
               ...
            ...
        except Exception as ex:
            self._logger.error(ex)
            msg = traceback.format_exc()
            self._logger.error('\n' + msg)
           ...
        finally:
            pass
        return res_model 
```
  

Oracle database connection pool init like this:
```
db_pool = PooledDB(cx_Oracle, threaded=True, user=xx,password=xx,dsn='den'
                   mincached=1, maxcached=1, maxconnections=500)

```

After this had done ,  the block problem that mentioned above has solved.  
**BUT**, new problem occurred : 
When I run a load test , the middleware does't performance very well .  With  thread pool size is 500 and database pool size is 500 , theoretically, about 500 of concurrency should be handled easily. 
But not . Actually, under  300 of concurrency ,  availability wasn't that high. I used siege to ran load test . command  liked this:
`$siege/bin/siege -c 300 -t 5M -f url.txt`

Looked in to the log file , I found 3 kinds of error and a 'LoopExit: This operation would block forever' error :
 ```
[2016-09-19 15:57:22,464] DEBUG::(7130 46912614902672)::BaseServices[line:136] - before, thread is 46912614902672 , thread_pool_maxsize is 500, thread_pool_size is 19
[2016-09-19 15:57:22,465] DEBUG::(7130 46912614902672)::BaseServices[line:138] - before, thread is 46912614902672, db_pool_maxsize is 500, db_pool_size is 44
[2016-09-19 15:57:26,270] DEBUG::(7130 46912614902672)::BaseServices[line:148] - after, p = None ,thread = 46912614902672, thread_pool_maxsize is 500, thread_pool_size is 19
[2016-09-19 15:57:26,270] DEBUG::(7130 46912614902672)::BaseServices[line:150] - after, thread = 46912614902672, db_pool_maxsize is 500, db_pool_size is 60
[2016-09-19 15:57:26,270] ERROR::(7130 46912614902672)::BaseServices[line:168] - 'NoneType' object has no attribute '__getitem__'
[2016-09-19 15:57:26,271] ERROR::(7130 46912614902672)::BaseServices[line:170] - 
Traceback (most recent call last):
  File "/home/was/python_apps/lcy_test/AsyncQueryProject/Services/BaseServices.py", line 151, in query_force_db
    for out_value in p[len(procedure_in_params):]:
TypeError: 'NoneType' object has no attribute '__getitem__'



[2016-09-19 15:57:29,337] DEBUG::(7130 46912765986288)::BaseServices[line:136] - before, thread is 46912765986288 , thread_pool_maxsize is 500, thread_pool_size is 21
[2016-09-19 15:57:29,337] DEBUG::(7130 46912765986288)::BaseServices[line:138] - before, thread is 46912765986288, db_pool_maxsize is 500, db_pool_size is 75
[2016-09-19 15:57:29,340] DEBUG::(7130 46912614793296)::BaseServices[line:34] - inside, parent thread is 46912765986288 ,thread is 46912614793296
[2016-09-19 15:57:30,348] ERROR::(7130 46912614793296)::BaseServices[line:41] - This operation would block forever
[2016-09-19 15:57:30,349] ERROR::(7130 46912614793296)::BaseServices[line:44] - 
inside exception, parent thread is 46912765986288 ,thread is 46912614793296 Traceback (most recent call last):
  File "/home/was/python_apps/lcy_test/AsyncQueryProject/Services/BaseServices.py", line 38, in do_db_operation
    % (op_id, thread.get_ident(), str(p), conn._pool._maxconnections, conn._pool._connections))
  File "/home/was/software/python2.7.8/lib/python2.7/logging/__init__.py", line 1148, in debug
    self._log(DEBUG, msg, args, **kwargs)
  File "/home/was/software/python2.7.8/lib/python2.7/logging/__init__.py", line 1279, in _log
    self.handle(record)
  File "/home/was/software/python2.7.8/lib/python2.7/logging/__init__.py", line 1289, in handle
    self.callHandlers(record)
  File "/home/was/software/python2.7.8/lib/python2.7/logging/__init__.py", line 1329, in callHandlers
    hdlr.handle(record)
  File "/home/was/software/python2.7.8/lib/python2.7/logging/__init__.py", line 755, in handle
    self.acquire()
  File "/home/was/software/python2.7.8/lib/python2.7/logging/__init__.py", line 706, in acquire
    self.lock.acquire()
  File "/home/was/software/python2.7.8/lib/python2.7/threading.py", line 173, in acquire
    rc = self.__block.acquire(blocking)
  File "_semaphore.pyx", line 112, in gevent._semaphore.Semaphore.acquire (gevent/gevent._semaphore.c:3004)
  File "/home/was/software/python2.7.8/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/hub.py", line 331, in switch
    return greenlet.switch(self)
LoopExit: This operation would block forever
[2016-09-19 15:57:30,349] DEBUG::(7130 46912614793296)::BaseServices[line:47] - inside end ,parent thread is 46912765986288 , thread is 46912614793296, thread_pool_maxsize is 500, thread_pool_size is 21
[2016-09-19 15:57:32,043] DEBUG::(7130 46912765986288)::BaseServices[line:148] - after, p = [u'1', 'success'] ,thread = 46912765986288, thread_pool_maxsize is 500, thread_pool_size is 21
[2016-09-19 15:57:32,043] DEBUG::(7130 46912765986288)::BaseServices[line:150] - after, thread = 46912765986288, db_pool_maxsize is 500, db_pool_size is 81
[2016-09-19 15:57:32,049] DEBUG::(7130 46912765986288)::BaseServices[line:178] - finally, thread = 46912765986288, db_pool_maxsize is 500, db_pool_size is 80




[2016-09-19 15:57:30,130] DEBUG::(7130 46912766021712)::BaseServices[line:136] - before, thread is 46912766021712 , thread_pool_maxsize is 500, thread_pool_size is 21
[2016-09-19 15:57:30,130] DEBUG::(7130 46912766021712)::BaseServices[line:138] - before, thread is 46912766021712, db_pool_maxsize is 500, db_pool_size is 91
[2016-09-19 15:57:30,133] DEBUG::(7130 46912615041296)::BaseServices[line:34] - inside, parent thread is 46912766021712 ,thread is 46912615041296
[2016-09-19 15:57:32,151] DEBUG::(7130 46912766021712)::BaseServices[line:148] - after, p = None ,thread = 46912766021712, thread_pool_maxsize is 500, thread_pool_size is 21
[2016-09-19 15:57:32,151] DEBUG::(7130 46912766021712)::BaseServices[line:150] - after, thread = 46912766021712, db_pool_maxsize is 500, db_pool_size is 65
[2016-09-19 15:57:32,151] ERROR::(7130 46912766021712)::BaseServices[line:168] - 'NoneType' object has no attribute '__getitem__'
[2016-09-19 15:57:32,152] ERROR::(7130 46912766021712)::BaseServices[line:170] - 
Traceback (most recent call last):
  File "/home/was/python_apps/lcy_test/AsyncQueryProject/Services/BaseServices.py", line 151, in query_force_db
    for out_value in p[len(procedure_in_params):]:
TypeError: 'NoneType' object has no attribute '__getitem__'
[2016-09-19 15:57:32,162] DEBUG::(7130 46912766021712)::BaseServices[line:178] - finally, thread = 46912766021712, db_pool_maxsize is 500, db_pool_size is 64




[2016-09-19 15:57:36,627] DEBUG::(7130 46912770525776)::BaseServices[line:136] - before, thread is 46912770525776 , thread_pool_maxsize is 500, thread_pool_size is 21
[2016-09-19 15:57:36,627] DEBUG::(7130 46912770525776)::BaseServices[line:138] - before, thread is 46912770525776, db_pool_maxsize is 500, db_pool_size is 133
[2016-09-19 15:57:36,630] DEBUG::(7130 46912615041776)::BaseServices[line:34] - inside, parent thread is 46912770525776 ,thread is 46912615041776
[2016-09-19 15:57:37,637] DEBUG::(7130 46912615041776)::BaseServices[line:38] - inside after, parent thread is 46912770525776 ,thread is 46912615041776, p is [u'1', 'success'], db_pool_maxsize is 500, db_pool_size is 123
[2016-09-19 15:57:38,088] DEBUG::(7130 46912770525776)::BaseServices[line:148] - after, p = None ,thread = 46912770525776, thread_pool_maxsize is 500, thread_pool_size is 21
[2016-09-19 15:57:38,088] DEBUG::(7130 46912770525776)::BaseServices[line:150] - after, thread = 46912770525776, db_pool_maxsize is 500, db_pool_size is 48
[2016-09-19 15:57:38,088] ERROR::(7130 46912770525776)::BaseServices[line:168] - 'NoneType' object has no attribute '__getitem__'
[2016-09-19 15:57:38,088] ERROR::(7130 46912770525776)::BaseServices[line:170] - 
Traceback (most recent call last):
  File "/home/was/python_apps/lcy_test/AsyncQueryProject/Services/BaseServices.py", line 151, in query_force_db
    for out_value in p[len(procedure_in_params):]:
TypeError: 'NoneType' object has no attribute '__getitem__'
[2016-09-19 15:57:38,092] DEBUG::(7130 46912770525776)::BaseServices[line:178] - finally, thread = 46912770525776, db_pool_maxsize is 500, db_pool_size is 47
```

It seems that : 
 **1.**  gevent thread pool doesn't gives a thread object immediately or a None object ?  

 **2.**  `p = c.callproc(procedure_name, procedure_params)` , this line , oracle doesn't return ?
     Oracle database allows more than 2500 connections which means the middleware's database pool size is more than enough. 

 **3.**  thread pool does give a thread object and oracle returned successfully, but gevent failed to match each return ? 
    I have run a script to prove whether each request and response matches each other .  The answer is yes .  The script takes 4 process and each process sends 300 post requests.  each request post a random int and response return the random int . accroding the script's log , each request and response matches each other very well except middleware occurs error . 

**4.**  the error : `LoopExit: This operation would block forever` , what kind of situation will call this error ? Does this error effect in this situation ?

Can anybody tell me why all of this happen ? That will be very very very grateful !!! 

By the way , when the concurrency is not that high , the middleware performance very well . 
You can  search the log's detail information  in attach file.  

[3 kinds of error.txt](https://github.com/gevent/gevent/files/484656/3.kinds.of.error.txt)
[original log.txt.zip](https://github.com/gevent/gevent/files/484693/original.log.txt.zip)

**Comment by jamadden** [comment]
Can you produce a [short, self-contained example](http://sscce.org) demonstrating the problem and eliminating the use of oracle? That's the best way to get something like this debugged.

Python isn't known for its multi-threading capability; performance will definitely suffer as the number of threads climbs.

A `LoopExit` like that is typically caused when you attempt a gevent blocking operation, but there are no other greenlets available to switch to. That's not a terribly surprising outcome to me in a threadpool scenario like this.

**Comment by vamkaxifa** [comment]
Hi jamadden 
Thanks for your help !
I finally completed the demo. This demo eliminated the use of oracle.
When you run the demo. You can make a request like this:
```
url : http://127.0.0.1:8800/p/pq
Content-Type : application/json
body: {"pra":4}
method: POST
```
The demo performances well , but when I run a load test , the performance go down and NoneType error occurs . Please refer to attachments for more information .  
Thank you very very very much !

[demo.zip](https://github.com/gevent/gevent/files/487010/demo.zip)
[8800.zip](https://github.com/gevent/gevent/files/487014/8800.zip)

**Comment by jamadden** [comment]
I'm afraid that's still far too complex. A self-contained example means it shouldn't have any third-party dependencies like `tornado` and ideally is a single file. 

I can generate load myself, you don't need to supply that part of it, but you do need to tell me what load level you see problems with; ideally for simulating load you'd use a standard tool like [ab](https://httpd.apache.org/docs/2.4/programs/ab.html).

**Comment by vamkaxifa** [comment]
Hi jamadden, This time I simplify the demo which don't have any third-party dependencies.
Attachment is a single file which is what you want. 
Sorry for I didn't use [ab](https://httpd.apache.org/docs/2.4/programs/ab.html) to simulate load , but I do use [seige](https://www.joedog.org/siege-home/) to simulate.
siege command : 
```
$HOME/siege/bin/siege -c 2 -t 5M -f url.txt > res.out

```
The command means use 2 threads to test the demo for 5 minutes . In my log, I found error log like this:
```
[2016-09-26 13:44:43,164] DEBUG::(24574 47682824084688)::demoMain[line:138] - after, p = None ,thread = 47682824084688, thread_pool_maxsize is 500, thread_pool_size is 2
[2016-09-26 13:44:43,165] ERROR::(24574 47682824084688)::demoMain[line:141] - unsupported operand type(s) for -: 'NoneType' and 'int'
[2016-09-26 13:44:43,165] ERROR::(24574 47682824084688)::demoMain[line:143] - 
Traceback (most recent call last):
  File "/home/was/python_apps/lcy_test/demoMain.py", line 139, in fun
    results = p - 1
TypeError: unsupported operand type(s) for -: 'NoneType' and 'int'

[2016-09-26 13:44:43,165] ERROR::(24574 47682824084688)::demoMain[line:166] - local variable 'results' referenced before assignment
[2016-09-26 13:44:43,166] ERROR::(24574 47682824084688)::demoMain[line:168] - 
Traceback (most recent call last):
  File "/home/was/python_apps/lcy_test/demoMain.py", line 159, in do_something
    res = base_services.fun(_data)
  File "/home/was/python_apps/lcy_test/demoMain.py", line 146, in fun
    return results
UnboundLocalError: local variable 'results' referenced before assignment
```
If you can't find the same error when you simulating load, you can use command below: 
```
$HOME/siege/bin/siege -c 500 -t 5M -f url.txt > res.out
```
[demoMain.py.zip](https://github.com/gevent/gevent/files/492338/demoMain.py.zip)

**Comment by jamadden** [comment]
There was still a lot of simplification that could be done. Here's the script I ended up testing:
```python
# coding:utf-8
from gevent import monkey
monkey.patch_all()

import json
import os
import signal
import time

from gevent.threadpool import ThreadPool
from gevent import thread
import gevent.wsgi
import gevent
from gevent import Timeout

class BaseServices:
    def __init__(self):
        self._db_handle_pool = ThreadPool(500)

    def do_db_operation(self, op_id, pra):
        time.sleep(0.2)
        ret = pra + 1
        return ret

    def fun(self, data):
        pra = data['pra']
        p = self._db_handle_pool.apply(self.do_db_operation, (str(thread.get_ident()), pra))
        assert p == pra + 1
        results = p - 1
        return results

base_services = BaseServices()

def application(env, start_response):
    if env['REQUEST_METHOD'] == 'POST':
        data = json.loads(env['wsgi.input'].read())
    else:
        data = {'pra': 1}

    with Timeout(3):
        base_services.fun(data)

    start_response('200 OK', [('Content-Type', 'text/html')])
    return [b'OK']


if __name__ == "__main__":
    server = gevent.wsgi.WSGIServer(('', 8800), application, log=None)

    def shutdown(*args):
        server.stop(timeout=5)
        os.exit(signal.SIGTERM)
    gevent.signal(signal.SIGTERM, shutdown)

    server.serve_forever()
```

Note that I parameterized whether to read the input based on the request type, *just in case* that made a difference (initially I was testing with GET requests and later added POST requests when I couldn't reproduce any problems).

**I can't reproduce the issue** you're seeing using either Python 3.4.5 with gevent master or Python 2.7.12 using gevent 1.1.2. I've tested both GET and POST requests. If you can still reproduce a problem even with this further simplification (errors would be very apparently printed to stderr and result in 500 responses, and your load testing tool should at least tell you how many of those there were), then it would appear to be related to your system.

To test POST requests, I issued this command:
```
$ ab -t 300 -c 520  -p /tmp/post.json -s 60 -r http://127.0.0.1:8800/
```

where `-t 300` says to run for up to five minutes using 50,000 requests, `-c 520` says to have 520 requests in flight at once (I verified that the `ThreadPool` did expand to a maximum of 500 threads), `-s 60` is the socket timeout, and `-r` says to ignore the occasional socket receive error (introduced, I believe, by kernel throttling). The file `/tmp/post.json` is the POST data, and has the contents `{"pra": 1}`.

On my machine that takes about 95s to run, and I'v run it repeatedly against the same python server process, and I've even run it in two different terminals at once (when I do that, it runs for more than 200s). Using `-k` to pipeline the requests with HTTP Keep-Alive doesn't make a difference. Adjusting the argument to `time.sleep` doesn't make a difference, other than making the test take longer (unless you go too small, and then the threadpool doesn't fill up).

At 500 threads, Python spends most of its time fighting over the GIL, it seems. The throughput is very "bursty," hence the addition of the increased socket timeout.

I get the same results with `siege -c 520 -t 300S -q -b  http://localhost:8800/`

**Comment by vamkaxifa** [comment]
Hi jamadden, I tested your script on my machine and it turned out that you were right, no error occurred. The only difference between your script and my script is that I use logger but you don't . For further testing , I add logger to your script : 
```
# coding:utf-8
from gevent import monkey
monkey.patch_all()

import json
import os
import signal
import time

from gevent.threadpool import ThreadPool
from gevent import thread
import gevent.wsgi
import gevent
from gevent import Timeout

import logging

logging.basicConfig(level=10)
logger = logging.getLogger()


class BaseServices:
    def __init__(self, _logger):
        self._db_handle_pool = ThreadPool(500)
        self._logger = _logger

    def do_db_operation(self, op_id, par):
        self._logger.debug('here is 111')
        time.sleep(0.2)
        ret = pra + 1
        self._logger.debug('here is 222 and ret is %i ' % ret)
        return ret

    def fun(self, data):
        pra = data['pra']
        self._logger.debug('here is 11')
        p = self._db_handle_pool.apply(self.do_db_operation, (str(thread.get_ident()), par))
        self._logger.debug('here is 22')
        assert p == pra + 1
        results = p - 1
        return results

base_services = BaseServices(logger)


def application(env, start_response):
    if env['REQUEST_METHOD'] == 'POST':
        data = json.loads(env['wsgi.input'].read())
    else:
        data = {'pra': 1}

    with Timeout(3):
        logger.debug('here is 1 ')
        base_services.fun(data)
        logger.debug('here is 2 ')

    start_response('200 OK', [('Content-Type', 'text/html')])
    return [b'OK']


if __name__ == "__main__":
    server = gevent.wsgi.WSGIServer(('', 8800), application, log=None)

    def shutdown(*args):
        server.stop(timeout=5)
        os.exit(signal.SIGTERM)
    gevent.signal(signal.SIGTERM, shutdown)

    server.serve_forever()
```
Then, test POST request: 
 `siege -c 520 -t 300S -f url.txt > res.out`
The file url.txt is the POST data, and has the contents `http://127.0.0.1:8800 POST {"pra":5}` .
The test resulted in failure:
```
[was@testint01 etc]$ $HOME/siege -c 520 -t 5M -f url.txt > res.out 
** SIEGE 3.0.8
** Preparing 520 concurrent users for battle.
The server is now under siege...
done.
siege aborted due to excessive socket failure; you
can change the failure threshold in $HOME/.siegerc

Transactions:		         497 hits
Availability:		       25.58 %
Elapsed time:		       10.18 secs
Data transferred:	        0.03 MB
Response time:		        6.01 secs
Transaction rate:	       48.82 trans/sec
Throughput:		        0.00 MB/sec
Concurrency:		      293.40
Successful transactions:         497
Failed transactions:	        1446
Longest transaction:	        3.77
Shortest transaction:	        0.00
```
In stdout, `AssertionError` could be found:
```
DEBUG:root:here is 22
error: cannot switch to a different thread
Traceback (most recent call last):
(<built-in method switch of Greenlet object at 0x2aaaab16c190>,
 <gevent._semaphore.Semaphore object at 0x2aecbf7ed090>) failed with error

  File "/home/was/software/python2.7.8/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/pywsgi.py", line 508, in handle_one_response
Traceback (most recent call last):
  File "_semaphore.pyx", line 50, in gevent._semaphore.Semaphore._notify_links (gevent/gevent._semaphore.c:1653)
    self.run_application()
  File "/home/was/software/python2.7.8/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/pywsgi.py", line 494, in run_application
    self.result = self.application(self.environ, self.start_response)
  File "/home/was/python_apps/lcy_test/jamadden_log.py", line 54, in application
    base_services.fun(data)
  File "/home/was/python_apps/lcy_test/jamadden_log.py", line 39, in fun
    assert p == pra + 1
AssertionError
{'CONTENT_LENGTH': '10',
 'CONTENT_TYPE': 'application/x-www-form-urlencoded',
 'GATEWAY_INTERFACE': 'CGI/1.1',
 'HTTP_ACCEPT': '*/*',
 'HTTP_ACCEPT_ENCODING': 'gzip',
 'HTTP_CONNECTION': 'close',
 'HTTP_HOST': '127.0.0.1:8800',
 'HTTP_USER_AGENT': 'Mozilla/5.0 (unknown-x86_64-linux-gnu) Siege/3.0.8',
 'PATH_INFO': '/p',
 'QUERY_STRING': '',
 'REMOTE_ADDR': '127.0.0.1',
 'REMOTE_PORT': '52950',
 'REQUEST_METHOD': 'POST',
 'SCRIPT_NAME': '',
 'SERVER_NAME': 'testint01',
 'SERVER_PORT': '8800',
 'SERVER_PROTOCOL': 'HTTP/1.1',
 'SERVER_SOFTWARE': 'gevent/1.0 Python/2.7',
 'wsgi.errors': <open file '<stderr>', mode 'w' at 0x2aecb8f201e0>,
 'wsgi.input': <gevent.pywsgi.Input object at 0x2aaaab534350>,
 'wsgi.multiprocess': False,
 'wsgi.multithread': False,
 'wsgi.run_once': False,
 'wsgi.url_scheme': 'http',
 'wsgi.version': (1, 0)} failed with AssertionError
```
In addition, there were also many `LoopExit: This operation would block forever` errors.

**Comment by jamadden** [comment]
Excellent, I was hoping you would spot that! That's how you create a SSCE: you strip away *everything* that doesn't cause a problem and you add things back in one at a time until you do have a problem. You want the simplest possible way to reproduce something.

So what does logging do that causes this problem? This is at least [hinted at](http://www.gevent.org/gevent.pywsgi.html#gevent.pywsgi.LoggingLogAdapter) in [some places](http://www.gevent.org/changelog.html#b1-jul-17-2015), and maybe it's explicitly documented somewhere, but logging acquires a lock in order to ensure its messages are serialized.  This modified (simplified 😄 ) code will result in the same errors:

```python
...
from gevent import lock
lock = lock.RLock()

class BaseServices:
    def __init__(self):
        self._db_handle_pool = ThreadPool(500)

    def do_db_operation(self, op_id, pra):
        with lock:
            gevent.sleep()
            ret = pra + 1
            return ret
    ...
```

In a monkey-patched system, you can only acquire locks if there is some other running greenlet to switch to, and in a threadpool, there typically are *no* other running greenlets---your code is the only thing that's running. Hence the `LoopExit` errors for an un-contended lock. In addition, like most objects in gevent (e.g., sockets), when they are in use by one thread, they can't be used by any other threads (because the other threads have different Hubs), thus the `error: cannot switch to a different thread` for a contended lock. 

Basically, you have to be extremely careful not to share gevent-using resources between threads, and logging implicitly does that behind your back. This is all a bit obscure, I admit, but it's a consequence of using multiple native threads.

In this particular case, you can get around it by declining to monkey-patch `thread`: `monkey.patch_all(thread=False)`.

**Comment by vamkaxifa** [comment]
I see,  so it is!  That explains!  All of this helps a lot!  I declined to monkey-patch `monkey.patch_all(thread=False)` in my middleware, turned off logger at the same time,  as a result , middleware performances much more better than before at same load level. That's excellent! Everything seems ok now! 
By the way, there was a error that appers frequently when load level is much too high in middleware. But it seems that this error doesn't matter considering that load level probably will not be that high in real case. But I insist on posting here, may be it matters for you 😄 or me in future 😄
```
Traceback (most recent call last):
  File "/home/was/software/python2.7.9/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/pywsgi.py", line 508, in handle_one_response
    self.run_application()
  File "/home/was/software/python2.7.9/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/pywsgi.py", line 495, in run_application
    self.process_result()
  File "/home/was/software/python2.7.9/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/pywsgi.py", line 486, in process_result
    self.write(data)
  File "/home/was/software/python2.7.9/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/pywsgi.py", line 376, in write
    self._write(data)
  File "/home/was/software/python2.7.9/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/pywsgi.py", line 369, in _write
    self._sendall(data)
  File "/home/was/software/python2.7.9/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/pywsgi.py", line 355, in _sendall
    self.socket.sendall(data)
  File "/home/was/software/python2.7.9/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/socket.py", line 458, in sandal
    data_sent += self.send(_get_memory(data, data_sent), flags)
  File "/home/was/software/python2.7.9/lib/python2.7/site-packages/gevent-1.0.1-py2.7-linux-x86_64.egg/gevent/socket.py", line 435, in send
    return sock.send(data, flags)
error: [Errno 32] Broken pipe
{'CONTENT_LENGTH': '295',
 'CONTENT_TYPE': 'application/x-www-form-urlencoded',
 'GATEWAY_INTERFACE': 'CGI/1.1',
 'HTTP_ACCEPT': '*/*',
 'HTTP_ACCEPT_ENCODING': 'gzip',
 'HTTP_CONNECTION': 'close',
 'HTTP_HOST': '127.0.0.1:8800',
 'HTTP_USER_AGENT': 'Mozilla/5.0 (unknown-x86_64-linux-gnu) Siege/3.0.8',
 'PATH_INFO': '/p',
 'QUERY_STRING': '',
 'REMOTE_ADDR': '127.0.0.1',
 'REMOTE_PORT': '45670',
 'REQUEST_METHOD': 'POST',
 'SCRIPT_NAME': '',
 'SERVER_NAME': 'testint01',
 'SERVER_PORT': '8800',
 'SERVER_PROTOCOL': 'HTTP/1.1',
 'SERVER_SOFTWARE': 'gevent/1.0 Python/2.7',
 'wsgi.errors': <open file '<stderr>', mode 'w' at 0x2aff4b4d51e0>,
 'wsgi.input': <gevent.pywsgi.Input object at 0x2aaab3698a10>,
 'wsgi.multiprocess': False,
 'wsgi.multithread': False,
 'wsgi.run_once': False,
 'wsgi.url_scheme': 'http',
 'wsgi.version': (1, 0)} failed with error
```

哥们，你太牛了，虽然我还没理清深层次的原因，但问题基本解决了。
来China的时候，记得来北京，请你吃下馆子吃烤鸭！

**Comment by jamadden** [comment]
A `Broken pipe` error there means the client has been disconnected. That's not uncommon during load testing.

This traceback is coming from the very old gevent 1.0. [gevent 1.1a1](http://www.gevent.org/changelog.html#a1-jun-29-2015) and newer [suppress this traceback](https://github.com/gevent/gevent/pull/377).

**Comment by advance512** [comment]
If you're facing a similar issue, #890 might be relevant, better give it a read.
